realm/realm-library/src/main/java/io/realm/sync/permissions/Permission.java

```diff
@@ -0,0 +1,562 @@
+/*
+ * Copyright 2018 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.sync.permissions;
+
+import io.realm.RealmObject;
+import io.realm.annotations.RealmClass;
+import io.realm.internal.annotations.ObjectServer;
+
+/**
+ * This class encapsulates the privileges granted a given {@link Role}. These privileges can be
+ * applied to either the entire Realm, Classes or individual objects.
+ * <p>
+ * If no privileges are defined for an individual object, the values for {@link ClassPermissions}
+ * will be inherited, if no values are defined there, the ones from {@link RealmPermissions} will
+ * be used. If no values can be found there, no privileges are granted.
```

Comment:
How would you describe it then? I assume the issue is that the high-level privileges might map to other low-level privileges? But not sure how to describe that accurately

Reply:
They don't commute at all. You can't change objects in any way (Update, Delete, SetPermissions, etc.) in a Realm that you don't have Update permissions to change. The individual privilege type does not "trickle down". This would quickly create some nonsensical situations that would make things hard to model.